src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -365,16 +369,93 @@ private static JsonElement convertSnakeCaseToCamelCaseRecursively(JsonElement js
 
     }
 
+    public static void verifyAndUpdateIntrospectRefreshTokenPayload(Main main, AppIdentifier appIdentifier,
+            Storage storage, JsonObject payload, String iss, String refreshToken) throws StorageQueryException, TenantOrAppNotFoundException, FeatureNotEnabledException, InvalidConfigException, IOException {
+        
+        OAuthStorage oauthStorage = StorageUtils.getOAuthStorage(storage);
+
+        if (!payload.get("active").getAsBoolean()) {
+            return; // refresh token is not active
+        }
+
+        Transformations.transformExt(payload);
+
+        long issuedAt = payload.get("iat").getAsLong();
+        String subject = payload.get("sub").getAsString();
+        String clientId = payload.get("client_id").getAsString();
+        String sessionHandle = null;
+        if (payload.has("sessionHandle")) {
+            sessionHandle = payload.get("sessionHandle").getAsString();
+        }
+
+        boolean isSubjectValid = !oauthStorage.isRevoked(appIdentifier, "sub", subject, issuedAt);
+        boolean isClientIdSubjectValid = !oauthStorage.isRevoked(appIdentifier, "client_id_sub", clientId + ":" + subject, issuedAt);
+        boolean isSessionHandleValid = true;
+        if (sessionHandle != null) {
```

Comment:
I think we should check these in a single query if at all possible.

Reply:
Done